File: py/src/jobshield/wage/radar.py
"""Wage Radar: per-occupation underpayment signal (spec 6.2).

Regress actual median wage on centrality over the whole market, then compare
the predicted wage (what the occupation "should" earn given its network
position) to its actual median. Positive gap = underpayment signal.
"""
from __future__ import annotations

from jobshield.types import (
    Centrality,
    OccupationCode,
    UnderpaymentSignal,
    WageStats,
)
from jobshield.wage.regression import regression_predict


def underpayment_signal(
    occ: OccupationCode,
    centrality: Centrality,
    wage_data: dict[OccupationCode, WageStats],
    all_occupations: list[OccupationCode],
) -> UnderpaymentSignal:
    """Compute (predicted, actual, gap_ratio) for `occ`.

    `centrality` must have both `degree` and (optionally) `betweenness`. We
    use degree centrality for the regression — it's the must-have, and
    spec 6.2 says Brandes is the stretch goal that we *also* compute, but
    the regression itself can run on either; degree is the simpler and
    more interpretable predictor.
    """
    xs = [centrality.degree[o] for o in all_occupations]
    ys = [wage_data[o].median for o in all_occupations]
    actual = wage_data[occ].median
    predicted = regression_predict(centrality.degree[occ], xs, ys)
    gap = (predicted - actual) / predicted if predicted != 0 else 0.0
    return UnderpaymentSignal(
        occ=occ,
        actual_wage=actual,
        predicted_wage=predicted,
        gap_ratio=gap,
    )


def all_underpayment_signals(
    centrality: Centrality,
    wage_data: dict[OccupationCode, WageStats],
) -> list[UnderpaymentSignal]:
    occs = list(wage_data.keys())
    return [underpayment_signal(o, centrality, wage_data, occs) for o in occs]


__all__ = ["all_underpayment_signals", "underpayment_signal"]

File: ts/api/src/radar.ts
// Compute per-occupation underpayment signal (matches Python radar.py).

import type {
  CentralityPayload,
  PipelineArtifacts,
  WageRadarRow,
} from "@jobshield/shared/contracts";

function regressionPredict(x: number, allX: number[], allY: number[]): number {
  const n = allX.length;
  if (n === 0 || allX.length !== allY.length) return 0;
  if (n < 2) {
    return allY.reduce((a, b) => a + b, 0) / n;
  }
  const meanX = allX.reduce((a, b) => a + b, 0) / n;
  const meanY = allY.reduce((a, b) => a + b, 0) / n;
  let sxx = 0;
  let sxy = 0;
  for (let i = 0; i < n; i++) {
    const dx = (allX[i] ?? 0) - meanX;
    sxx += dx * dx;
    sxy += dx * ((allY[i] ?? 0) - meanY);
  }
  if (sxx === 0) return meanY;
  const slope = sxy / sxx;
  return slope * (x - meanX) + meanY;
}

export function computeWageRadar(artifacts: PipelineArtifacts): WageRadarRow[] {
  const occs = Object.keys(artifacts.wage_data);
  const xs = occs.map((o) => artifacts.centrality.degree[o] ?? 0);
  const ys = occs.map((o) => artifacts.wage_data[o]?.median ?? 0);
  return occs.map((occ) => {
    const x = artifacts.centrality.degree[occ] ?? 0;
    const y = artifacts.wage_data[occ]?.median ?? 0;
    const predicted = regressionPredict(x, xs, ys);
    const gap = predicted !== 0 ? (predicted - y) / predicted : 0;
    return {
      occ,
      label: occ,
      centrality: x,
      wage: y,
      underpaid: gap > 0.05, // >5% gap = underpaid signal (positive gap = underpaid)
      gap_ratio: gap,
    };
  });
}
